### DISCUSSION TITLE
Adding constraints on the manipulator and the external axis

### DISCUSSION THREAD

**Discussion by mohamedsayed18** [tesseract-robotics/tesseract_planning#139] [Q&A] [answered]
I'm trying to make the tool of the robot always point down and the external axis should rotate to compensate for this.
 I thought about the `cartesian_coeff` but as I understand this will add cost/constraint to the generated path between two waypoints, not adding constraints on the robot itself.
 I think the right choice is the `joint_coeff` by adding a high coefficient value for the joints I don't want to move a lot, but that seems like not working. 
 Any help would be appreciated, Note: I'm working on a 3D printing application.

**Comment by Levi-Armstrong** [comment, marked as answer]
In this case since the default trajopt plan profile does not provide what you need I would create a new profile. I would inherit form the TrajOptDefaultPlanProfile and override the apply method for the cartesian waypoint. In this function I would do exactly what it is doing so it adds the dynamic cartesian constraint or cost. Then add an additional cartesian constraint using the world frame as the desired frame and then set the coefficients to [0,0,0,5,5,0] which will create a constraint that it must be oriented with world but allowed to rotate  around the world z-axis. You could also take it a step further if you are using descartes and create a custom sampler which only generates sampled states which satisfy this to provided a better seed to trajopt. I would only do this if trajopt has issues finding a solution.

**Reply by mohamedsayed18** [reply]
I'll work on this, thank you so much.

**Reply by mohamedsayed18** [reply]
I made a draft PR that contain a profile planner. https://github.com/tesseract-robotics/tesseract_planning/pull/145
I'm still not able to get the desired behavior. Talking about one plan instruction for simplicity I noticed that the generated path for one instruction is subjected to these constraints but what I want to achieve is to make the final point(plan instruction) also subjected to the same constraints (have a specific orientation).

**Reply by Levi-Armstrong** [reply]
Is the waypoint definition a Cartesian waypoint or Joint waypoint?

**Reply by mohamedsayed18** [reply]
Cartesian waypoint
I think if it is defined as a joint waypoint it will achieve the joint values at the end of the instruction, the constraint will have no effect.

**Reply by Levi-Armstrong** [reply]
Ok, yea that is correct. If the plan instruction contains a Cartesian waypoint then that should also have the Cartesian constraints but you are seeing that all intermediate states do but the last way points does not?

**Reply by mohamedsayed18** [reply]
No, the intermediate points are a smooth transition from the start pose to the end pose.

**Reply by mohamedsayed18** [reply]
Do the costs/constraint on rotation mean don't rotate with respect to a specific frame? For example, if I'm moving between two points that have the same orientation but this orientation is different from the world frame where we have rotation constraint, In this case as I understand the constraint condition will be achieved as we don't need rotation between the two points.

What if I want to align the rotation of my target with a specific frame, considering we are having an external axis to achieve this.

**Reply by mohamedsayed18** [reply]
I was thinking about using the `joint_coeff` to restrict joints in the robot to make the tool always point down. for example, make the wrist joint always between -45 to 45.
As I understand `joint_coeff` is only applied in joint waypoints or state waypoints. I tried to add `joint_coeff` in the `glass_upright_example` but it doesn't have any effect. is `joint_coeff` implemented?
Another question, Is it applicable to add `joint_coeff` for a cartesian waypoint?

**Reply by Levi-Armstrong** [reply]
The joint_coeff is for the robot joint position constraints/cost.

> As I understand joint_coeff is only applied in joint waypoints or state waypoints. I tried to add joint_coeff in the glass_upright_example but it doesn't have any effect. is joint_coeff implemented?

That is correct.

> Another question, Is it applicable to add joint_coeff for a cartesian waypoint?

It is, but the default profile does not support it. The default profiles are a general example that may work for some use cases. In this case I would say you would be better off creating your own profile using the default as a guide which adds the necessary costs/constraints based on your profiles configuration settings.

**Reply by mohamedsayed18** [reply]
To constraint the targets to a specific alignment I was thinking about limiting the joint value of the end effector between two values.
can I use the `upper_tols` and `lower_tols` to achieve this?

or is it better to use a different method to achieve the targets' alignment?

**Reply by Levi-Armstrong** [reply]
The tolerance functionality in TrajOpt has had little testing and not sure it works as expected. If you plan to leverage tolerances I recommend using TrajOptIfopt.

**Reply by mohamedsayed18** [reply]
I did several trials with joint tolerance but it seems like not the right tool for my problem.
Regarding using the trajopt profile(I made a new one inherited from trajopt) to add a constraint to my problem, it restricts the waypoint from rotation around the `world` frame but it didn't restrict the waypoint to have a specific orientation.
In the following gif, I have the target with the z-axis pointing parallel to the external axis plate. this is the output without adding any constraints.
![No_constraints](https://user-images.githubusercontent.com/27130471/146197258-18518c20-74ee-4655-8641-ccfe942dd366.gif)

To add the constraint I added `createCartesianWaypointTermInfo`, with `world` as the working frame and  the coefficients are [0,0,0,5,5,0]
![Rot_constrained](https://user-images.githubusercontent.com/27130471/146197147-46be57fb-d235-42f6-a93c-7c21236dafbd.gif)

as shown, the rotation of the waypoint is constrained with respect to the `world`. But what I was trying to achieve was to constraint the target to have a specific orientation(z-axis pointing down).
As an overview of the problem I'm trying to solve. The input is a cartesian waypoint (position and orientation with respect to the external axis). If the waypoint's z-axis is not pointing down with respect to the `world` frame, the external axis should change its orientation to achieve the desired orientation.

**Reply by Levi-Armstrong** [reply]
is the last waypoint in the trajectory the waypoint you are wanting to constrain? If that is the case then the positioner should rotate 90 degrees so it would be point down? Also, can you point me to what you currently have?

**Reply by mohamedsayed18** [reply]
>is the last waypoint in the trajectory the waypoint you are wanting to constrain? 

Yes, but the bigger goal is to make even the intermediate points constrained. considering this motion as continuous printing.
>If that is the case then the positioner should rotate 90 degrees so it would be point down?

Yes, that is how I think we could keep the tool always pointing down.

>can you point me to what you currently have?

[Here](https://github.com/mohamedsayed18/tesseract_planning/blob/875c91e2e9019ede92b33bf12497196802a0100c/tesseract_motion_planners/trajopt/src/profile/trajopt_constraint_plan_profile.cpp) I inherited the trajopt default profile and added an extra constraint for the cartesian waypoint.
But right now I think of creating a new TermInfo for example `CreateAlignTermInfo` which aligns the cartesian waypoint with the `world`, but I'm still thinking if it is possible to do so?
